src/os/FileStore.cc

```diff
@@ -2889,69 +2889,142 @@ int FileStore::fiemap(coll_t cid, const ghobject_t& oid,
 {
   tracepoint(objectstore, fiemap_enter, cid.c_str(), offset, len);
 
-  if (!backend->has_fiemap() || len <= (size_t)m_filestore_fiemap_threshold) {
+  if (len <= (size_t)m_filestore_fiemap_threshold) {
     map<uint64_t, uint64_t> m;
     m[offset] = len;
     ::encode(m, bl);
     return 0;
   }
 
-
-  struct fiemap *fiemap = NULL;
-  map<uint64_t, uint64_t> exomap;
-
   dout(15) << "fiemap " << cid << "/" << oid << " " << offset << "~" << len << dendl;
 
+  map<uint64_t, uint64_t> exomap;
+  bool fallback_to_fiemap = false;
   FDRef fd;
+  struct fiemap *fiemap = NULL;
+  uint64_t i;
+  struct fiemap_extent *extent = NULL;
+
   int r = lfn_open(cid, oid, false, &fd);
   if (r < 0) {
     dout(10) << "read couldn't open " << cid << "/" << oid << ": " << cpp_strerror(r) << dendl;
-  } else {
-    uint64_t i;
+    goto done;
+  }
 
-    r = backend->do_fiemap(**fd, offset, len, &fiemap);
-    if (r < 0)
+#if defined(__linux__) && defined(SEEK_HOLE) && defined(SEEK_DATA)
+  off_t hole_pos, data_pos;
+
+  // If compiled on an OS with SEEK_HOLE/SEEK_DATA support, but running
+  // on an OS that doesn't support SEEK_HOLE/SEEK_DATA, EINVAL is returned.
+  // Fall back to use fiemap.
+  hole_pos = lseek(**fd, 0, SEEK_HOLE);
+  if (hole_pos < 0) {
+    if (errno == EINVAL) {
+      fallback_to_fiemap = true;
+    } else {
+      r = -errno;
       goto done;
+    }
+  }
 
-    if (fiemap->fm_mapped_extents == 0) {
-      free(fiemap);
+  // Seek back to pos 0 first if there is a hole.
+  if (hole_pos > 0) {
+    off_t tmp_pos;
+    tmp_pos = lseek(**fd, 0, SEEK_SET);
+    if (tmp_pos != 0) {
+      r = -errno;
       goto done;
     }
+  }
+
+  if (!fallback_to_fiemap) {
+    // If lseek fails with errno setting to be ENXIO, this means the current
```

Comment:
From the man page of lseek, the error of ENXIO means: whence is SEEK_DATA or SEEK_HOLE, and the current file offset is beyond the end of the file. So I think it's ENXIO here.